File: addons/terrabrush/Scripts/TerrainControlDock.cs
#if TOOLS
using Godot;
using System.IO;
using System.Linq;

namespace TerraBrush;

[Tool]
public partial class TerrainControlDock : Control {
    private int _selectedBrushIndex = 0;
    private TerrainToolType _selectedTool = TerrainToolType.TerrainAdd;
    private TerrainToolType _temporaryTool = TerrainToolType.None;
    private int? _selectedTextureIndex = null;
    private int? _selectedFoliageIndex = null;
    private int? _selectedObjectIndex = null;

    [NodePath] private FlowContainer _brushesContainer;
    [NodePath] private FlowContainer _toolTypesContainer;
    [NodePath] private FlowContainer _texturesContainer;
    [NodePath] private FlowContainer _foliagesContainer;
    [NodePath] private FlowContainer _objectsContainer;
    [NodePath] private Slider _brushSizeSlider;
    [NodePath] private Slider _brushStrengthSlider;

    public TerraBrush TerraBrush { get;set; }
    public BrushDecal BrushDecal { get;set; }
    public EditorResourcePreview EditorResourcePreview { get;set; }

    public override void _Ready() {
        base._Ready();
        this.RegisterNodePaths();

        if (TerraBrush != null) {
            _brushSizeSlider.Value = TerraBrush.BrushSize;
            _brushStrengthSlider.Value = TerraBrush.BrushStrength;
            _selectedTool = TerraBrush.TerrainTool;
            _selectedBrushIndex = TerraBrush.SelectedBrushIndex.GetValueOrDefault();
            _selectedTextureIndex = TerraBrush.TextureSetIndex;
            _selectedFoliageIndex = TerraBrush.FoliageIndex;
            _selectedObjectIndex = TerraBrush.ObjectIndex;

            InitializeBrushes();
            InitializeToolPreview();
            InitializeTextures();
            InitializeFoliages();
            InitializeObjects();
        }

        _brushSizeSlider.ValueChanged += value => {
            SetBrushSize((int) value);
        };

        _brushStrengthSlider.ValueChanged += value => {
            SetBrushStrength((float) value);
        };
    }

    public override void _ExitTree() {
        base._ExitTree();
    }

    private void InitializeBrushes() {
        CustomContentLoader.AddBrushesPreviewToParent(_brushesContainer, index => {
            SetSelectedBrushIndex(index);
        });

        UpdateSelectedBrush();
    }

    public void SetBrushSize(int value) {
        _brushSizeSlider.Value = value;
        TerraBrush.SetBrushSize(value);
        BrushDecal.SetSize(value);
    }

    public void SetBrushStrength(float value) {
        _brushStrengthSlider.Value = value;
        TerraBrush.SetBrushStrength(value);
    }

    public void SetSelectedBrushIndex(int index) {
        _selectedBrushIndex = index;

        UpdateSelectedBrush();
    }

    private void UpdateSelectedBrush() {
        var brushesPreview = _brushesContainer.GetChildren();

        for (var i = 0; i < brushesPreview.Count; i++) {
            var brushPreview = brushesPreview[i];

            ((DockPreviewButton) brushPreview).ButtonPressed = i == _selectedBrushIndex;
        }

        var brushImage = ((DockPreviewButton) brushesPreview[_selectedBrushIndex]).ButtonImage.GetImage();
        TerraBrush?.SetCurrentBrush(_selectedBrushIndex, brushImage);
        BrushDecal?.SetBrushImage(brushImage);
    }

    private void InitializeToolPreview() {
        foreach (ToolPreview toolPreview in _toolTypesContainer.GetChildren()) {
            toolPreview.OnSelect = () => {
                SelectToolType(toolPreview.ToolType);
            };
        }

        UpdateSelectedTerrainTool();
    }

    public void SelectToolType(TerrainToolType toolType) {
        _selectedTool = toolType;
        UpdateSelectedTerrainTool();
        TerraBrush.SetTerrainTool(_selectedTool);
    }

    private void UpdateSelectedTerrainTool() {
        foreach (ToolPreview toolPreview in _toolTypesContainer.GetChildren()) {
            toolPreview.ButtonPressed = toolPreview.ToolType == (_temporaryTool == TerrainToolType.None ? _selectedTool : _temporaryTool);
        }
    }

    private void InitializeTextures() {
        CustomContentLoader.AddTexturesPreviewToParent(TerraBrush, _texturesContainer, index => {
            SetSelectedTextureIndex(index);
        });

        var textureSets = TerraBrush?.TextureSets?.TextureSets;
        if (textureSets != null) {
            foreach (var texturePreview in _texturesContainer.GetChildren().OfType<DockPreviewButton>()) {
                int index = texturePreview.GetIndex();
                if (index < 0 || index >= textureSets.Length) {
                    continue;
                }

                var textureSet = textureSets[index];
                texturePreview.TooltipText = textureSet.Name ?? $"Texture {index + 1}";
            }
        }

        if (_texturesContainer.GetChildCount() > 0 && _selectedTextureIndex == null) {
            _selectedTextureIndex = 0;
        }
        UpateSelectedTextureSet();
    }

    public void SetSelectedTextureIndex(int index) {
        _selectedTextureIndex = index;

        UpateSelectedTextureSet();
    }

    private void UpateSelectedTextureSet() {
        var texturesPreview = _texturesContainer.GetChildren();

        for (var i = 0; i < texturesPreview.Count; i++) {
            var texturePreview = texturesPreview[i];

            ((DockPreviewButton) texturePreview).ButtonPressed = i == _selectedTextureIndex;
        }

        TerraBrush?.SetTextureSet(_selectedTextureIndex);
    }

    private void InitializeFoliages() {
        CustomContentLoader.AddFoliagesPreviewToParent(TerraBrush, _foliagesContainer, index => {
            SetSelectedFoliageIndex(index);
        });

        var foliages = TerraBrush?.Foliages;
        if (foliages != null) {

            foreach (var foliagePreview in _foliagesContainer.GetChildren().OfType<DockPreviewButton>()) {
                int index = foliagePreview.GetIndex();
                if (index < 0 || index >= foliages.Length) {
                    continue;
                }

                var foliage = foliages[index];
                string tooltipText = $"Foliage {index + 1}";

                var meshPath = foliage?.Definition?.Mesh?.ResourcePath;
                if (!string.IsNullOrEmpty(meshPath)) {
                    tooltipText = Path.GetFileName(meshPath);
                }

                foliagePreview.TooltipText = tooltipText;
            }
        }

        if (TerraBrush.Foliages?.Length > 0 && _selectedFoliageIndex == null) {
            _selectedFoliageIndex = 0;
        }

        UpdateSelectedFoliage();
    }

    public void SetSelectedFoliageIndex(int index) {
        _selectedFoliageIndex = index;

        UpdateSelectedFoliage();
    }

    private void UpdateSelectedFoliage() {
        var foliagesPreview = _foliagesContainer.GetChildren();

        for (var i = 0; i < foliagesPreview.Count; i++) {
            var foliagePreview = foliagesPreview[i];

            ((DockPreviewButton) foliagePreview).ButtonPressed = i == _selectedFoliageIndex;
        }

        TerraBrush?.SetFoliage(_selectedFoliageIndex);
    }

    private void InitializeObjects() {
        CustomContentLoader.AddObjectsPreviewToParent(TerraBrush, _objectsContainer, index => {
            SetSelectedObjectIndex(index);
        });

        var objects = TerraBrush?.Objects;
        if (objects != null) {
            foreach (var objectPreview in _objectsContainer.GetChildren().OfType<DockPreviewButton>()) {
                int index = objectPreview.GetIndex();
                if (index < 0 || index >= objects.Length) {
                    continue;
                }

                var objectItem = objects[index];
                string tooltipText = $"Object {index + 1}";

                var packedScenes = objectItem?.Definition?.ObjectScenes;
                if (packedScenes?.Length > 0) {
                    var sceneNames = packedScenes
                        .Where(scene => scene != null)
                        .Select(scene => Path.GetFileName(scene.ResourcePath))
                        .Where(name => !string.IsNullOrEmpty(name));

                    var joinedNames = string.Join(", ", sceneNames);
                    if (!string.IsNullOrEmpty(joinedNames)) {
                        tooltipText = joinedNames;
                    }
                }

                objectPreview.TooltipText = tooltipText;
            }
        }

        if (TerraBrush.Objects?.Length > 0 && _selectedObjectIndex == null) {
            _selectedObjectIndex = 0;
        }

        UpdateSelectedObject();
    }

    public void SetSelectedObjectIndex(int index) {
        _selectedObjectIndex = index;

        UpdateSelectedObject();
    }

    private void UpdateSelectedObject() {
        var objectsPreview = _objectsContainer.GetChildren();

        for (var i = 0; i < objectsPreview.Count; i++) {
            var objectPreview = objectsPreview[i];

            ((DockPreviewButton) objectPreview).ButtonPressed = i == _selectedObjectIndex;
        }

        TerraBrush?.SetObject(_selectedObjectIndex);
    }

    public void SetShiftPressed(bool pressed) {
        if (pressed) {
            if (_selectedTool == TerrainToolType.TerrainAdd || _selectedTool == TerrainToolType.TerrainRemove || _selectedTool == TerrainToolType.TerrainFlatten) {
                _temporaryTool = TerrainToolType.TerrainSmooth;
            } else if (_selectedTool == TerrainToolType.FoliageAdd) {
                _temporaryTool = TerrainToolType.FoliageRemove;
            } else if (_selectedTool == TerrainToolType.ObjectAdd) {
                _temporaryTool = TerrainToolType.ObjectRemove;
            } else if (_selectedTool == TerrainToolType.WaterAdd) {
                _temporaryTool = TerrainToolType.WaterRemove;
            } else if (_selectedTool == TerrainToolType.WaterFlowAdd) {
                _temporaryTool = TerrainToolType.WaterFlowRemove;
            } else if (_selectedTool == TerrainToolType.SnowAdd) {
                _temporaryTool = TerrainToolType.SnowRemove;
            } else if (_selectedTool == TerrainToolType.HoleAdd) {
                _temporaryTool = TerrainToolType.HoleRemove;
            } else if (_selectedTool == TerrainToolType.LockAdd) {
                _temporaryTool = TerrainToolType.LockRemove;
            } else {
                _temporaryTool = TerrainToolType.None;
            }
        } else {
            _temporaryTool = TerrainToolType.None;
        }

        TerraBrush.SetTerrainTool(_temporaryTool == TerrainToolType.None ? _selectedTool : _temporaryTool);
        UpdateSelectedTerrainTool();
    }
}
#endif

File: src/editor_nodes/terrain_control_dock.cpp
#include "terrain_control_dock.h"

#include "tool_preview.h"
#include "dock_preview_button.h"
#include "../terra_brush.h"
#include "../misc/custom_content_loader.h"
#include "../misc/enums.h"

#include <godot_cpp/classes/margin_container.hpp>
#include <godot_cpp/classes/v_box_container.hpp>
#include <godot_cpp/classes/tab_container.hpp>
#include <godot_cpp/classes/control.hpp>
#include <godot_cpp/classes/scroll_container.hpp>
#include <godot_cpp/classes/h_flow_container.hpp>
#include <godot_cpp/classes/resource_loader.hpp>
#include <godot_cpp/classes/label.hpp>
#include <godot_cpp/classes/h_slider.hpp>

using namespace godot;

void TerrainControlDock::_bind_methods() {
    ClassDB::bind_method(D_METHOD("onBrushSizeValueChange", "value"), &TerrainControlDock::onBrushSizeValueChange);
    ClassDB::bind_method(D_METHOD("onBrushStrengthValueChange", "value"), &TerrainControlDock::onBrushStrengthValueChange);
    ClassDB::bind_method(D_METHOD("selectToolType", "toolType"), &TerrainControlDock::selectToolType);
    ClassDB::bind_method(D_METHOD("setSelectedBrushIndex", "index"), &TerrainControlDock::setSelectedBrushIndex);
    ClassDB::bind_method(D_METHOD("setSelectedTextureIndex", "index"), &TerrainControlDock::setSelectedTextureIndex);
    ClassDB::bind_method(D_METHOD("setSelectedFoliageIndex", "index"), &TerrainControlDock::setSelectedFoliageIndex);
    ClassDB::bind_method(D_METHOD("setSelectedObjectIndex", "index"), &TerrainControlDock::setSelectedObjectIndex);
    ClassDB::bind_method(D_METHOD("setSelectedMetaInfoIndex", "index"), &TerrainControlDock::setSelectedMetaInfoIndex);

    ADD_SIGNAL(MethodInfo("toolTypeSelected", PropertyInfo(Variant::INT, "toolType")));
    ADD_SIGNAL(MethodInfo("brushSelected", PropertyInfo(Variant::INT, "index")));
    ADD_SIGNAL(MethodInfo("brushSizeChanged", PropertyInfo(Variant::INT, "value")));
    ADD_SIGNAL(MethodInfo("brushStrengthChanged", PropertyInfo(Variant::FLOAT, "value")));
    ADD_SIGNAL(MethodInfo("textureSelected", PropertyInfo(Variant::INT, "index")));
    ADD_SIGNAL(MethodInfo("foliageSelected", PropertyInfo(Variant::INT, "index")));
    ADD_SIGNAL(MethodInfo("objectSelected", PropertyInfo(Variant::INT, "index")));
    ADD_SIGNAL(MethodInfo("metaInfoSelected", PropertyInfo(Variant::INT, "index")));
}

TerrainControlDock::TerrainControlDock() {
    _selectedBrushIndex = 0;
    _selectedTool = TerrainToolType::TERRAINTOOLTYPE_TERRAINADD;
    _selectedTextureIndex = -1;
    _selectedFoliageIndex = -1;
    _selectedObjectIndex = -1;
    _selectedMetaInfoIndex = -1;
}

TerrainControlDock::~TerrainControlDock() {}

void TerrainControlDock::_ready() {
    buildLayout();

    if (_terraBrush != nullptr) {
        initializeBrushes();
        initializeToolPreview();
        initializeTextures();
        initializeFoliages();
        initializeObjects();
        initializeMetaInfoLayers();
    }

    _brushSizeSlider->connect("value_changed", Callable(this, "onBrushSizeValueChange"));
    _brushStrengthSlider->connect("value_changed", Callable(this, "onBrushStrengthValueChange"));
}

void TerrainControlDock::set_terraBrush(TerraBrush *value) {
    _terraBrush = value;
}

void TerrainControlDock::set_editorResourcePreview(EditorResourcePreview *value) {
    _editorResourcePreview = value;
}

void TerrainControlDock::initializeBrushes() {
    CustomContentLoader::addBrushesPreviewToParent(_brushesContainer, Callable(this, "setSelectedBrushIndex"));

    updateSelectedBrush();
}

void TerrainControlDock::initializeToolPreview() {
    for (int i = 0; i < _toolTypesContainer->get_child_count(); i++) {
        Node *childNode = _toolTypesContainer->get_child(i);
        ToolPreview *toolPreview = Object::cast_to<ToolPreview>(childNode);
        toolPreview->set_onSelect(Callable(this, "selectToolType").bind(toolPreview->get_toolType()));
    }

    updateSelectedTerrainTool();
}

void TerrainControlDock::initializeTextures() {
    CustomContentLoader::addTexturesPreviewToParent(_terraBrush, _texturesContainer, Callable(this, "setSelectedTextureIndex"));
}

void TerrainControlDock::initializeFoliages() {
    CustomContentLoader::addFoliagesPreviewToParent(_terraBrush, _foliagesContainer, Callable(this, "setSelectedFoliageIndex"));
}

void TerrainControlDock::initializeObjects() {
    CustomContentLoader::addObjectsPreviewToParent(_terraBrush, _objectsContainer, Callable(this, "setSelectedObjectIndex"));
}

void TerrainControlDock::initializeMetaInfoLayers() {
    CustomContentLoader::addMetaInfoLayersPreviewToParent(_terraBrush, _metaInfoLayersContainer, Callable(this, "setSelectedMetaInfoIndex"));
}

void TerrainControlDock::updateSelectedBrush() {
    for (int i = 0; i < _brushesContainer->get_child_count(); i++) {
        Node *childNode = _brushesContainer->get_child(i);
        DockPreviewButton *brushPreview = Object::cast_to<DockPreviewButton>(childNode);

        brushPreview->set_pressed(i == _selectedBrushIndex);
    }

    Node *selectedNode = _brushesContainer->get_child(_selectedBrushIndex);
    DockPreviewButton *selectedDockButton = Object::cast_to<DockPreviewButton>(selectedNode);

    emit_signal("brushSelected", _selectedBrushIndex);
}

void TerrainControlDock::updateSelectedTerrainTool() {
    for (int i = 0; i < _toolTypesContainer->get_child_count(); i++) {
        Node *childNode = _toolTypesContainer->get_child(i);
        ToolPreview *toolPreview = Object::cast_to<ToolPreview>(childNode);
        toolPreview->set_pressed(toolPreview->get_toolType() == _selectedTool);
    }
}

void TerrainControlDock::upateSelectedTextureSet() {
    for (int i = 0; i < _texturesContainer->get_child_count(); i++) {
        Node *childNode = _texturesContainer->get_child(i);
        DockPreviewButton *previewButton = Object::cast_to<DockPreviewButton>(childNode);

        previewButton->set_pressed(i == _selectedTextureIndex);
    }

    emit_signal("textureSelected", _selectedTextureIndex);
}

void TerrainControlDock::updateSelectedFoliage() {
    for (int i = 0; i < _foliagesContainer->get_child_count(); i++) {
        Node *childNode = _foliagesContainer->get_child(i);
        DockPreviewButton *previewButton = Object::cast_to<DockPreviewButton>(childNode);

        previewButton->set_pressed(i == _selectedFoliageIndex);
    }

    emit_signal("foliageSelected", _selectedFoliageIndex);
}

void TerrainControlDock::updateSelectedObject() {
    for (int i = 0; i < _objectsContainer->get_child_count(); i++) {
        Node *childNode = _objectsContainer->get_child(i);
        DockPreviewButton *previewButton = Object::cast_to<DockPreviewButton>(childNode);

        previewButton->set_pressed(i == _selectedObjectIndex);
    }

    emit_signal("objectSelected", _selectedObjectIndex);
}

void TerrainControlDock::updateSelectedMetaInfo() {
    for (int i = 0; i < _metaInfoLayersContainer->get_child_count(); i++) {
        Node *childNode = _metaInfoLayersContainer->get_child(i);
        DockPreviewButton *previewButton = Object::cast_to<DockPreviewButton>(childNode);

        previewButton->set_pressed(i == _selectedMetaInfoIndex);
    }

    emit_signal("metaInfoSelected", _selectedMetaInfoIndex);
}

void TerrainControlDock::onBrushSizeValueChange(const float value) {
    setBrushSize((int)value);
}

void TerrainControlDock::onBrushStrengthValueChange(const float value) {
    setBrushStrength((float)value);
}

void TerrainControlDock::setBrushSize(int value) {
    _brushSizeSlider->set_value(value);
    emit_signal("brushSizeChanged", value);
}

void TerrainControlDock::setBrushStrength(float value) {
    _brushStrengthSlider->set_value(value);
    emit_signal("brushStrengthChanged", value);
}

void TerrainControlDock::setSelectedBrushIndex(const int index) {
    _selectedBrushIndex = index;

    updateSelectedBrush();
}

void TerrainControlDock::selectToolType(const TerrainToolType toolType) {
    _selectedTool = toolType;
    updateSelectedTerrainTool();
    emit_signal("toolTypeSelected", toolType);
}

void TerrainControlDock::setSelectedTextureIndex(const int index) {
    _selectedTextureIndex = index;

    upateSelectedTextureSet();
}

void TerrainControlDock::setSelectedFoliageIndex(const int index) {
    _selectedFoliageIndex = index;

    updateSelectedFoliage();
}

void TerrainControlDock::setSelectedObjectIndex(const int index) {
    _selectedObjectIndex = index;

    updateSelectedObject();
}

void TerrainControlDock::setSelectedMetaInfoIndex(const int index) {
    _selectedMetaInfoIndex = index;

    updateSelectedMetaInfo();
}

// This component has some complexe ui so I create a kinda of structure using {} for better visualization
void TerrainControlDock::buildLayout() {
    set_anchors_and_offsets_preset(LayoutPreset::PRESET_FULL_RECT);

    MarginContainer *marginContainer = memnew(MarginContainer);
    marginContainer->set_anchors_and_offsets_preset(LayoutPreset::PRESET_FULL_RECT);
    marginContainer->set((StringName)"theme_override_constants/margin_left", 5);
    marginContainer->set((StringName)"theme_override_constants/margin_top", 5);
    marginContainer->set((StringName)"theme_override_constants/margin_right", 5);
    marginContainer->set((StringName)"theme_override_constants/margin_bottom", 5);
    add_child(marginContainer);
    { // MarginContainer
        VBoxContainer *vBoxContainer = memnew(VBoxContainer);
        marginContainer->add_child(vBoxContainer);
        { // VBoxContainer
            TabContainer *tabContainer1 = memnew(TabContainer);
            tabContainer1->set_current_tab(0);
            tabContainer1->set_v_size_flags(SizeFlags::SIZE_EXPAND_FILL);
            vBoxContainer->add_child(tabContainer1);
            { // TabContainer
                Control *toolsTab = memnew(Control);
                toolsTab->set_name("Tools");
                tabContainer1->add_child(toolsTab);
                { //Tab
                    MarginContainer *toolsMarginContainer = memnew(MarginContainer);
                    toolsMarginContainer->set_anchors_and_offsets_preset(LayoutPreset::PRESET_FULL_RECT);
                    toolsMarginContainer->set((StringName)"theme_override_constants/margin_left", 10);
                    toolsMarginContainer->set((StringName)"theme_override_constants/margin_top", 10);
                    toolsMarginContainer->set((StringName)"theme_override_constants/margin_right", 10);
                    toolsMarginContainer->set((StringName)"theme_override_constants/margin_bottom", 10);
                    toolsTab->add_child(toolsMarginContainer);
                    { // MarginContainer
                        ScrollContainer *toolsScrollContainer = memnew(ScrollContainer);
                        toolsMarginContainer->add_child(toolsScrollContainer);
                        { // ScrollContainer
                            VBoxContainer *toolsVBoxContainer = memnew(VBoxContainer);
                            toolsVBoxContainer->set_h_size_flags(SizeFlags::SIZE_EXPAND_FILL);
                            toolsVBoxContainer->set_v_size_flags(SizeFlags::SIZE_EXPAND_FILL);
                            toolsScrollContainer->add_child(toolsVBoxContainer);
                            { // VBoxContainer
                                _toolTypesContainer = memnew(HFlowContainer);
                                toolsVBoxContainer->add_child(_toolTypesContainer);
                                { // HFlowContainer
                                    ToolPreview *toolPreview = memnew(ToolPreview);
                                    toolPreview->set_toolType(TerrainToolType::TERRAINTOOLTYPE_TERRAINADD);
                                    toolPreview->set_buttonImage(ResourceLoader::get_singleton()->load("res://addons/terrabrush/Assets/Icons/map_add.png"));
                                    toolPreview->set_tooltip_text("Raise terrain");
                                    _toolTypesContainer->add_child(toolPreview);

                                    toolPreview = memnew(ToolPreview);
                                    toolPreview->set_toolType(TerrainToolType::TERRAINTOOLTYPE_TERRAINREMOVE);
                                    toolPreview->set_buttonImage(ResourceLoader::get_singleton()->load("res://addons/terrabrush/Assets/Icons/map_remove.png"));
                                    toolPreview->set_tooltip_text("Lower terrain");
                                    _toolTypesContainer->add_child(toolPreview);

                                    toolPreview = memnew(ToolPreview);
                                    toolPreview->set_toolType(TerrainToolType::TERRAINTOOLTYPE_TERRAINSMOOTH);
                                    toolPreview->set_buttonImage(ResourceLoader::get_singleton()->load("res://addons/terrabrush/Assets/Icons/map_smooth.png"));
                                    toolPreview->set_tooltip_text("Smooth terrain");
                                    _toolTypesContainer->add_child(toolPreview);

                                    toolPreview = memnew(ToolPreview);
                                    toolPreview->set_toolType(TerrainToolType::TERRAINTOOLTYPE_TERRAINFLATTEN);
                                    toolPreview->set_buttonImage(ResourceLoader::get_singleton()->load("res://addons/terrabrush/Assets/Icons/map_flatten.png"));
                                    toolPreview->set_tooltip_text("Flatten terrain");
                                    _toolTypesContainer->add_child(toolPreview);

                                    toolPreview = memnew(ToolPreview);
                                    toolPreview->set_toolType(TerrainToolType::TERRAINTOOLTYPE_TERRAINSETHEIGHT);
                                    toolPreview->set_buttonImage(ResourceLoader::get_singleton()->load("res://addons/terrabrush/Assets/Icons/map_set_height.png"));
                                    toolPreview->set_tooltip_text("Set terrain height");
                                    _toolTypesContainer->add_child(toolPreview);

                                    toolPreview = memnew(ToolPreview);
                                    toolPreview->set_toolType(TerrainToolType::TERRAINTOOLTYPE_TERRAINSETANGLE);
                                    toolPreview->set_buttonImage(ResourceLoader::get_singleton()->load("res://addons/terrabrush/Assets/Icons/map_set_angle.png"));
                                    toolPreview->set_tooltip_text("Set terrain angle");
                                    _toolTypesContainer->add_child(toolPreview);

                                    toolPreview = memnew(ToolPreview);
                                    toolPreview->set_toolType(TerrainToolType::TERRAINTOOLTYPE_PAINT);
                                    toolPreview->set_buttonImage(ResourceLoader::get_singleton()->load("res://addons/terrabrush/Assets/Icons/paint.png"));
                                    toolPreview->set_tooltip_text("Paint texture");
                                    _toolTypesContainer->add_child(toolPreview);

                                    toolPreview = memnew(ToolPreview);
                                    toolPreview->set_toolType(TerrainToolType::TERRAINTOOLTYPE_FOLIAGEADD);
                                    toolPreview->set_buttonImage(ResourceLoader::get_singleton()->load("res://addons/terrabrush/Assets/Icons/foliage_add.png"));
                                    toolPreview->set_tooltip_text("Add foliage");
                                    _toolTypesContainer->add_child(toolPreview);

                                    toolPreview = memnew(ToolPreview);
                                    toolPreview->set_toolType(TerrainToolType::TERRAINTOOLTYPE_FOLIAGEREMOVE);
                                    toolPreview->set_buttonImage(ResourceLoader::get_singleton()->load("res://addons/terrabrush/Assets/Icons/foliage_remove.png"));
                                    toolPreview->set_tooltip_text("Remove foliage");
                                    _toolTypesContainer->add_child(toolPreview);

                                    toolPreview = memnew(ToolPreview);
                                    toolPreview->set_toolType(TerrainToolType::TERRAINTOOLTYPE_OBJECTADD);
                                    toolPreview->set_buttonImage(ResourceLoader::get_singleton()->load("res://addons/terrabrush/Assets/Icons/object_add.png"));
                                    toolPreview->set_tooltip_text("Add objects");
                                    _toolTypesContainer->add_child(toolPreview);

                                    toolPreview = memnew(ToolPreview);
                                    toolPreview->set_toolType(TerrainToolType::TERRAINTOOLTYPE_OBJECTREMOVE);
                                    toolPreview->set_buttonImage(ResourceLoader::get_singleton()->load("res://addons/terrabrush/Assets/Icons/object_remove.png"));
                                    toolPreview->set_tooltip_text("Remove objects");
                                    _toolTypesContainer->add_child(toolPreview);

                                    toolPreview = memnew(ToolPreview);
                                    toolPreview->set_toolType(TerrainToolType::TERRAINTOOLTYPE_WATERADD);
                                    toolPreview->set_buttonImage(ResourceLoader::get_singleton()->load("res://addons/terrabrush/Assets/Icons/water_add.png"));
                                    toolPreview->set_tooltip_text("Add water");
                                    _toolTypesContainer->add_child(toolPreview);

                                    toolPreview = memnew(ToolPreview);
                                    toolPreview->set_toolType(TerrainToolType::TERRAINTOOLTYPE_WATERREMOVE);
                                    toolPreview->set_buttonImage(ResourceLoader::get_singleton()->load("res://addons/terrabrush/Assets/Icons/water_remove.png"));
                                    toolPreview->set_tooltip_text("Remove water");
                                    _toolTypesContainer->add_child(toolPreview);

                                    toolPreview = memnew(ToolPreview);
                                    toolPreview->set_toolType(TerrainToolType::TERRAINTOOLTYPE_WATERFLOWADD);
                                    toolPreview->set_buttonImage(ResourceLoader::get_singleton()->load("res://addons/terrabrush/Assets/Icons/flow_add.png"));
                                    toolPreview->set_tooltip_text("Add water flow");
                                    _toolTypesContainer->add_child(toolPreview);

                                    toolPreview = memnew(ToolPreview);
                                    toolPreview->set_toolType(TerrainToolType::TERRAINTOOLTYPE_WATERFLOWREMOVE);
                                    toolPreview->set_buttonImage(ResourceLoader::get_singleton()->load("res://addons/terrabrush/Assets/Icons/flow_remove.png"));
                                    toolPreview->set_tooltip_text("Remove water flow");
                                    _toolTypesContainer->add_child(toolPreview);

                                    toolPreview = memnew(ToolPreview);
                                    toolPreview->set_toolType(TerrainToolType::TERRAINTOOLTYPE_SNOWADD);
                                    toolPreview->set_buttonImage(ResourceLoader::get_singleton()->load("res://addons/terrabrush/Assets/Icons/snow_add.png"));
                                    toolPreview->set_tooltip_text("Add snow");
                                    _toolTypesContainer->add_child(toolPreview);

                                    toolPreview = memnew(ToolPreview);
                                    toolPreview->set_toolType(TerrainToolType::TERRAINTOOLTYPE_SNOWREMOVE);
                                    toolPreview->set_buttonImage(ResourceLoader::get_singleton()->load("res://addons/terrabrush/Assets/Icons/snow_remove.png"));
                                    toolPreview->set_tooltip_text("Remove snow");
                                    _toolTypesContainer->add_child(toolPreview);

                                    toolPreview = memnew(ToolPreview);
                                    toolPreview->set_toolType(TerrainToolType::TERRAINTOOLTYPE_HOLEADD);
                                    toolPreview->set_buttonImage(ResourceLoader::get_singleton()->load("res://addons/terrabrush/Assets/Icons/holes_add.png"));
                                    toolPreview->set_tooltip_text("Add hole");
                                    _toolTypesContainer->add_child(toolPreview);

                                    toolPreview = memnew(ToolPreview);
                                    toolPreview->set_toolType(TerrainToolType::TERRAINTOOLTYPE_HOLEREMOVE);
                                    toolPreview->set_buttonImage(ResourceLoader::get_singleton()->load("res://addons/terrabrush/Assets/Icons/holes_remove.png"));
                                    toolPreview->set_tooltip_text("Remove hole");
                                    _toolTypesContainer->add_child(toolPreview);

                                    toolPreview = memnew(ToolPreview);
                                    toolPreview->set_toolType(TerrainToolType::TERRAINTOOLTYPE_LOCKADD);
                                    toolPreview->set_buttonImage(ResourceLoader::get_singleton()->load("res://addons/terrabrush/Assets/Icons/lock_add.png"));
                                    toolPreview->set_tooltip_text("Lock");
                                    _toolTypesContainer->add_child(toolPreview);

                                    toolPreview = memnew(ToolPreview);
                                    toolPreview->set_toolType(TerrainToolType::TERRAINTOOLTYPE_LOCKREMOVE);
                                    toolPreview->set_buttonImage(ResourceLoader::get_singleton()->load("res://addons/terrabrush/Assets/Icons/lock_remove.png"));
                                    toolPreview->set_tooltip_text("Unlock");
                                    _toolTypesContainer->add_child(toolPreview);

                                    toolPreview = memnew(ToolPreview);
                                    toolPreview->set_toolType(TerrainToolType::TERRAINTOOLTYPE_METAINFOADD);
                                    toolPreview->set_buttonImage(ResourceLoader::get_singleton()->load("res://addons/terrabrush/Assets/Icons/tag_add.png"));
                                    toolPreview->set_tooltip_text("Add meta info");
                                    _toolTypesContainer->add_child(toolPreview);

                                    toolPreview = memnew(ToolPreview);
                                    toolPreview->set_toolType(TerrainToolType::TERRAINTOOLTYPE_METAINFOREMOVE);
                                    toolPreview->set_buttonImage(ResourceLoader::get_singleton()->load("res://addons/terrabrush/Assets/Icons/tag_remove.png"));
                                    toolPreview->set_tooltip_text("Remove meta info");
                                    _toolTypesContainer->add_child(toolPreview);
                                }

                                VBoxContainer *toolsBrushSizeVBoxContainer = memnew(VBoxContainer);
                                toolsVBoxContainer->add_child(toolsBrushSizeVBoxContainer);
                                { // VBoxContainer
                                    Label *label = memnew(Label);
                                    label->set_text("Brush size");
                                    toolsBrushSizeVBoxContainer->add_child(label);

                                    _brushSizeSlider = memnew(HSlider);
                                    _brushSizeSlider->set_min(1.0);
                                    _brushSizeSlider->set_max(200.0);
                                    _brushSizeSlider->set_value(1.0);
                                    toolsBrushSizeVBoxContainer->add_child(_brushSizeSlider);
                                }

                                VBoxContainer *toolsBrushStrengthVBoxContainer = memnew(VBoxContainer);
                                toolsVBoxContainer->add_child(toolsBrushStrengthVBoxContainer);
                                { // VBoxContainer
                                    Label *label = memnew(Label);
                                    label->set_text("Brush strength");
                                    toolsBrushStrengthVBoxContainer->add_child(label);

                                    _brushStrengthSlider = memnew(HSlider);
                                    _brushStrengthSlider->set_min(0.01);
                                    _brushStrengthSlider->set_max(1.0);
                                    _brushStrengthSlider->set_value(0.01);
                                    _brushStrengthSlider->set_step(0.01);
                                    toolsBrushStrengthVBoxContainer->add_child(_brushStrengthSlider);
                                }

                                VBoxContainer *toolsBrushesVBoxContainer = memnew(VBoxContainer);
                                toolsVBoxContainer->add_child(toolsBrushesVBoxContainer);
                                { // VBoxContainer
                                    Label *label = memnew(Label);
                                    label->set_text("Brushes");
                                    toolsBrushesVBoxContainer->add_child(label);

                                    _brushesContainer = memnew(HFlowContainer);
                                    toolsBrushesVBoxContainer->add_child(_brushesContainer);
                                }
                            }
                        }
                    }
                }
            }

            TabContainer *tabContainer2 = memnew(TabContainer);
            tabContainer2->set_current_tab(0);
            tabContainer2->set_v_size_flags(SizeFlags::SIZE_EXPAND_FILL);
            vBoxContainer->add_child(tabContainer2);
            { // TabContainer
                Control *texturesTab = memnew(Control);
                texturesTab->set_name("Textures");
                tabContainer2->add_child(texturesTab);
                { // Tab
                    MarginContainer *texturesMarginContainer = memnew(MarginContainer);
                    texturesMarginContainer->set_anchors_and_offsets_preset(LayoutPreset::PRESET_FULL_RECT);
                    texturesTab->add_child(texturesMarginContainer);
                    { // MarginContainer
                        _texturesContainer = memnew(HFlowContainer);
                        texturesMarginContainer->add_child(_texturesContainer);
                    }
                }

                Control *foliagesTab = memnew(Control);
                foliagesTab->set_name("Foliages");
                tabContainer2->add_child(foliagesTab);
                { // Tab
                    MarginContainer *foliagesMarginContainer = memnew(MarginContainer);
                    foliagesMarginContainer->set_anchors_and_offsets_preset(LayoutPreset::PRESET_FULL_RECT);
                    foliagesTab->add_child(foliagesMarginContainer);
                    { // MarginContainer
                        _foliagesContainer = memnew(HFlowContainer);
                        foliagesMarginContainer->add_child(_foliagesContainer);
                    }
                }

                Control *objectsTab = memnew(Control);
                objectsTab->set_name("Objects");
                tabContainer2->add_child(objectsTab);
                { // Tab
                    MarginContainer *objectsMarginContainer = memnew(MarginContainer);
                    objectsMarginContainer->set_anchors_and_offsets_preset(LayoutPreset::PRESET_FULL_RECT);
                    objectsTab->add_child(objectsMarginContainer);
                    { // MarginContainer
                        _objectsContainer = memnew(HFlowContainer);
                        objectsMarginContainer->add_child(_objectsContainer);
                    }
                }

                Control *metaInfoTab = memnew(Control);
                metaInfoTab->set_name("MetaInfo");
                tabContainer2->add_child(metaInfoTab);
                { // Tab
                    MarginContainer *metaInfoMarginContainer = memnew(MarginContainer);
                    metaInfoMarginContainer->set_anchors_and_offsets_preset(LayoutPreset::PRESET_FULL_RECT);
                    metaInfoTab->add_child(metaInfoMarginContainer);
                    { // MarginContainer
                        _metaInfoLayersContainer = memnew(HFlowContainer);
                        metaInfoMarginContainer->add_child(_metaInfoLayersContainer);
                    }
                }
            }
        }
    }
}